docs_nnx/why.rst

```diff
@@ -0,0 +1,416 @@
+^^^^^^^^^^
+The first improvement is that Flax NNX modules are regular Python objects, so you can easily
+construct and inspect them. Because Flax Linen Modules are lazy some attributes are not available
+upon construction and are only accesible at runtime, this makes it hard to inspect and debug.
+
+.. codediff::
+  :title: Linen, NNX
+  :sync:
+
+  class Block(nn.Module):
+    def setup(self):
+      self.linear = nn.Dense(10)
+
+  block = Block()
+
+  try:
+    block.linear  # AttributeError: "Block" object has no attribute "linear".
+  except AttributeError as e:
+    pass
+
+
+
+
+
+  ...
+
+  ---
+
+  class Block(nnx.Module):
+    def __init__(self, rngs):
+      self.linear = nnx.Linear(5, 10, rngs=rngs)
+
+  block = Block(nnx.Rngs(0))
+
+
+  block.linear
+  # Linear(
+  #   kernel=Param(
+  #     value=Array(shape=(5, 10), dtype=float32)
+  #   ),
+  #   bias=Param(
+  #     value=Array(shape=(10,), dtype=float32)
+  #   ),
+  #   ...
+
+Notice that in Flax NNX there is no shape inference so both the input and output shapes must be provided
+to the Linear module. This is a tradeoff that allows for more explicit and predictable behavior.
+
+Running Computation
+^^^^^^^^^^^^^^^^^^^
+In Flax Linen, all top-level computation must be done through the ``init`` or ``apply`` methods and the
+parameters or any other type of state is handled as a separate structure. This creates an asymmetry
+between code that runs inside ``apply`` that can run methods and other Modules directly, and code
+outside of ``apply`` that must use the ``apply`` method. In Flax NNX, there's no special context
+as parameters are held as attributes and methods can be called directly.
+
+.. codediff::
+  :title: Linen, NNX
+  :sync:
+
+  Encoder = lambda: nn.Dense(10)
+  Decoder = lambda: nn.Dense(2)
+
+  class AutoEncoder(nn.Module):
+    def setup(self):
+      self.encoder = Encoder()
+      self.decoder = Decoder()
+
+    def __call__(self, x) -> jax.Array:
+      return self.decoder(self.encoder(x))
+
+    def encode(self, x) -> jax.Array:
+      return self.encoder(x)
+
+  x = jnp.ones((1, 2))
+  model = AutoEncoder()
+  params = model.init(random.key(0), x)['params']
+
+  y = model.apply({'params': params}, x)
+  z = model.apply({'params': params}, x, method='encode')
+  y = Decoder().apply({'params': params['decoder']}, z)
+
+  ---
+
+  Encoder = lambda rngs: nnx.Linear(2, 10, rngs=rngs)
+  Decoder = lambda rngs: nnx.Linear(10, 2, rngs=rngs)
+
+  class AutoEncoder(nnx.Module):
+    def __init__(self, rngs):
+      self.encoder = Encoder(rngs)
+      self.decoder = Decoder(rngs)
+
+    def __call__(self, x) -> jax.Array:
+      return self.decoder(self.encoder(x))
+
+    def encode(self, x) -> jax.Array:
+      return self.encoder(x)
+
+  x = jnp.ones((1, 2))
+  model = AutoEncoder(nnx.Rngs(0))
+
+
+  y = model(x)
+  z = model.encode(x)
+  y = model.decoder(z)
+
+Note that in Linen, calling submodules directly is might not be possible as they are not initialized
```

Comment:
> Note that in Linen, calling submodules directly is might not be possible as they are not initialized so you must construct a new instance, and also the proper parameter structure must be provided.

"Note that in Linen, calling submodules directly is not possible as they are not initialized. So you must construct a new instance and provide proper parameter structure."